bin/crew

```diff
@@ -556,77 +561,114 @@ def download
-          FileUtils.cp cachefile, CREW_BREW_DIR, verbose: @fileutils_verbose
-          puts "Archive found in cache".lightgreen
-          return {source: source, filename: filename}
+
+    case File.basename(filename)
+    # Sources that download with curl
+    when /\.zip$/i, /\.(tar(\.(gz|bz2|xz|lz))?|tgz|tbz|txz)$/i, /\.deb$/i
+      # Recall file from cache if requeseted
+      if CREW_IS_CACHING
+        cachefile = CREW_CACHE_DIR + filename
+        if File.file?(cachefile)
+          if Digest::SHA256.hexdigest( File.read(cachefile) ) == sha256sum then
+            FileUtils.cp cachefile, CREW_BREW_DIR, verbose: @fileutils_verbose
+            puts "Archive found in cache".lightgreen
+            return {source: source, filename: filename}
           else
-          puts 'Cached archive checksum mismatch. :/ Will download.'.lightred
+            puts 'Cached archive checksum mismatch. :/ Will download.'.lightred
+          end
+        puts 'Cannot find cached archive. :/ Will download.'.lightred
         end
       end
-    end
-
-    system "#{CURL} --retry 3 -#{@verbose}#LC - --insecure \'#{url}\' --output #{filename}"
-
-    abort 'Checksum mismatch. :/ Try again.'.lightred unless
-      Digest::SHA256.hexdigest( File.read(filename) ) == sha256sum
-    puts "Archive downloaded".lightgreen
+      # Download file if not cached
+      system "#{CURL} --retry 3 -#{@verbose}#LC - --insecure \'#{url}\' --output #{filename}"
+      abort 'Checksum mismatch. :/ Try again.'.lightred unless
+        Digest::SHA256.hexdigest( File.read(filename) ) == sha256sum
+      puts "Archive downloaded".lightgreen
+      # Stow file in cache if requested
+      if CREW_IS_CACHING
+        puts 'Caching downloaded archive'.lightgreen
+        FileUtils.cp filename, CREW_CACHE_DIR, verbose: @fileutils_verbose
+      end
 
-    if ENV['CREW_CACHE_OPT']
-      puts 'Caching downloaded archive'.lightgreen
-      FileUtils.cp filename, CREW_CACHE_DIR, verbose: @fileutils_verbose
+    # Sources that download with git
+    when /\.git$/i
+      # Recall repository from cache if requested
+      if CREW_IS_CACHING
+        cachefile = CREW_CACHE_DIR + '/' + filename + '.tar.xz'
+        if File.file?(cachefile)
+          if system "sha256sum -c #{cachefile}.sha256"
+            FileUtils.mkdir @extract_dir
+            system "tar x#{@verbose}f #{cachefile} -C #{@extract_dir}"
+            return {source: source, filename: filename}
+          else
+            puts 'Cached repository checksum mismatch. :/ Will download.'.lightred
+          end
+        else
+          puts 'Cannot find cached repository. :/ Will download.'.lightred
+        end
+      end
+      # Download file if not needed
+      Dir.mkdir @extract_dir
+      Dir.chdir @extract_dir do
+        system 'git init'
+        system 'git config advice.detachedHead false'
+        system 'git config init.defaultBranch master'
```

Comment:
This is not necessarily a good assumption.  Not all repos default to master and some repos don't even have a master branch.

Reply:
No, but it doesn't matter anyway. This is just to suppress a warning. Git won't error or cause a problem if you set the default branch to master and the repository doesn't have a master branch.

Reply:
Then it spits out a concerning warning.

Reply:
If you specified master and that branch didn't exist, wouldn't you still get an error?

Reply:
Nope!